src/shared/vscode/commands2.ts

```diff
@@ -365,44 +369,192 @@ interface CommandInfo<T extends Callback> {
+     * The indexes of args in `execute()` that will be used to determine
+     * the uniqueness of the metric `vscode_executeCommand`.
+     * 
+     * - By default, `vscode_executeCommand` is throttled for metrics of the
+     *   same `id`.
+     * - Defining arg indexes for this property, the values of the args will be combined
+     *   with the `id`. This ensures throttling only happens for the same `id` + args
+     *   instead.
+     * - Indexes starts at 0.
+     *
+     * @example
+     * const argIndexes: ArgIndexes = {
+     *     0: undefined // the value is not important
+     *     1: 'source' // the value is the "source" attribute of `vscode_executeCommand`
+     * }
+     * ...
+     * myCommand.execute('thisIsUsed', 'SOURCE', 'thisIsNotUsed')
+     * ...
+     * // vscode_executeCommand -> { command: 'myCommand' source: 'SOURCE' } 
+     */
+    readonly argIndexes?: ArgIndexes
+}
 
+/** 
+ * Used by {@link CommandInfo}, will indicate that the arg should be used
+ * as a telemetry attribute.
+ */
+export type ArgIndexes = { [index: number]: ArgIndexType }
+type ArgIndexType = MetricAttribute | undefined
+
+/** Supported `vscode_executeCommand` attributes */
+const MetricAttributes = {
+    source: 'source'
+} as const
+type MetricAttribute = typeof MetricAttributes[keyof typeof MetricAttributes]
+
+function getInstrumenter(
+    id: { id: string; args: any[]; argIndexes: ArgIndexes },
+    threshold: number,
+    telemetryName?: MetricName
+) {
+    const currentTime = globals.clock.Date.now()
+    const info = TelemetryDebounceInfo.instance.get(id)
+
+    // to reduce # of events actually emitted, we throttle the same* event for brief period of time
+    // and instead increment a counter for executions while throttled.
     if (!telemetryName && info?.startTime !== undefined && currentTime - info.startTime < threshold) {
         info.debounceCount += 1
-        telemetryInfo.set(id, info)
-        getLogger().debug(`commands: skipped telemetry for "${id}"`)
+        TelemetryDebounceInfo.instance.set(id, info)
+        getLogger().debug(`commands: skipped telemetry for "${id.id}"`)
 
         return undefined
     }
 
     // Throttling occurs regardless of whether or not the instrumenter is invoked
     const span = telemetryName ? telemetry[telemetryName] : telemetry.vscode_executeCommand
     const debounceCount = info?.debounceCount !== 0 ? info?.debounceCount : undefined
-    telemetryInfo.set(id, { startTime: currentTime, debounceCount: 0 })
+    TelemetryDebounceInfo.instance.set(id, { startTime: currentTime, debounceCount: 0 })
+
+    const attributes = findAttributesToAddToMetric(id.args, id.argIndexes)
 
     return <T extends Callback>(fn: T, ...args: Parameters<T>) =>
         span.run(span => {
-            ;(span as Metric<VscodeExecuteCommand>).record({ command: id, debounceCount })
+            ;(span as Metric<VscodeExecuteCommand>).record({
+                command: id.id,
+                debounceCount,
```

Comment:
off topic for this PR, but 

- we should probably rename `debounceCount` to `collapsedCount` or something more intuitive.
- we should also include the compositeKey in the metric